apps/shopify/shopify.star

```diff
@@ -0,0 +1,554 @@
+        total_price = total_price + (float(order["subtotal_price"]) - excluded_line_item_price)
+    total_price = int(total_price)
+    print("${} total line item price".format(total_price))
+    return total_price
+
+def get_line_items_from_orders(orders, excluded_skus):
+    line_items = [li for li in flatten([o["line_items"] for o in orders]) if should_line_item_be_excluded(li, excluded_skus) == False]
+    return line_items
+
+def get_total_units(orders, excluded_skus):
+    line_items = [li for li in flatten([o["line_items"] for o in orders]) if should_line_item_be_excluded(li, excluded_skus) == False]
+    total = 0
+    for li in line_items:
+        total = total + li["quantity"]
+    return int(total)
+
+def get_total_units_plot_data(orders, excluded_skus, timezone):
+    if len(orders) == 0:
+        return None
+    max_hour = date_from_shopify_date(orders[-1]["created_at"]).in_location(timezone).hour
+    hourly_units = [0 for e in range(max_hour + 1)]
+    for order in orders:
+        order_date = date_from_shopify_date(order["created_at"]).in_location(timezone)
+        order_hour = order_date.hour
+        total_units = 0
+        for li in order["line_items"]:
+            if should_line_item_be_excluded(li, excluded_skus) == False:
+                total_units = int(li["quantity"]) + total_units
+        hourly_units[order_hour] = hourly_units[order_hour] + total_units
+    hourly_units = [(idx, units) for idx, units in enumerate(hourly_units)]
+    return hourly_units
+
+def get_total_orders_plot_data(orders, excluded_skus, timezone):
+    if len(orders) == 0:
+        return None
+    max_hour = date_from_shopify_date(orders[-1]["created_at"]).in_location(timezone).hour
+    hourly_orders = [0 for e in range(max_hour + 1)]
+    for order in orders:
+        order_date = date_from_shopify_date(order["created_at"]).in_location(timezone)
+        order_hour = order_date.hour
+        if should_order_be_excluded(order, excluded_skus) == False:
+            hourly_orders[order_hour] = hourly_orders[order_hour] + 1
+    hourly_orders = [(idx, units) for idx, units in enumerate(hourly_orders)]
+    return hourly_orders
+
+def should_order_be_excluded(order, excluded_skus):
+    lines = [li for li in order["line_items"] if should_line_item_be_excluded(li, excluded_skus) == False]
+    return len(lines) == 0
+
+def should_line_item_be_excluded(line_item, excluded_skus):
+    if line_item["gift_card"]:
+        return True
+
+    if excluded_skus:
+        sku = line_item["sku"]
+        for ex_sku in excluded_skus:
+            if ex_sku in sku:
+                # print("{} is excluded".format(sku))
+                return True
+
+    # print("{} is not excluded".format(sku))
+
+    return False
+
+def cache_orders(orders, start_time, store_name, api_token):
+    cache_key = "orders_{}_{}_{}".format(start_time, store_name, api_token[-6:] if len(api_token) >= 1 else "")
```

Comment:
It would be good to use the full `api_token` here. Limiting ourselves to the last 6 characters makes it easier for an attacker to guess a token suffix and access the cached order data. The ideal might be to compute a hash of the full `api_token` and `store_name` and use that as cache key.